### ISSUE TITLE
Oracle decode function. Difficult/impossible? to write the code myself.

### ISSUE THREAD

**Issue by GoogleCodeExporter** [sagarswathi/h2database#353]
```
Oracle decode function has variable number of parameters, and the parameter 
type may also vary (DATE, NUMBER, VARCHAR2). I have tried:

public final static Value decode(Value exp, Value ... params).

That works in many cases. But it fails in these cases:
a) 2 - decode(...)
b) sum(decode(...))

Decode is very much used in Oracle applications, and it is a big job to change 
all decode to 'case when ...'.

Decode should be supported directly by H2, or a guideline for how to write the 
decode self should be given.
```

Original issue reported on code.google.com by `SvRe...@gmail.com` on 17 Oct 2011 at 7:30

**Comment by GoogleCodeExporter** [comment]
```
I guess the easiest solution is to add it to H2, in org.h2.expression.Function. 
Patches are welcome of course :-)
```

Original comment by `thomas.t...@gmail.com` on 17 Oct 2011 at 7:00

**Comment by GoogleCodeExporter** [comment]
```
I have made a patch for decode in version 1.3.160. It is in the Class Parser. 
Here I parse the 'decode(...)' in a new method 'readDecodeWhen()' (similar to 
readWhen), and create a CASEWHEN function as if it was a CASE WHEN ... 
construction. See the followin lines in attached file.
187
2465 - 2466
2674 - 2711

I have tested it with different count of parameters, and different types. Also 
tested the cases above (2-decode, sum(decode)).
```

Original comment by `SvRe...@gmail.com` on 19 Oct 2011 at 1:31

**Comment by GoogleCodeExporter** [comment]
```
A bug in my Parser patch:
select decode(null, null, null, 5.0) returned 5.0 instead of null.
A fix to to this is to change Comparison.EQUAL to Comparison.EQUAL_NULL_SAFE
in readDecodeWhen() (line 2701).
See Attached file.
```

Original comment by `SvRe...@gmail.com` on 31 Oct 2011 at 12:53

**Comment by GoogleCodeExporter** [comment]
```
Some JUnit tests of decode.
```

Original comment by `SvRe...@gmail.com` on 31 Oct 2011 at 8:29

**Comment by GoogleCodeExporter** [comment]
```
This is now committed in the patch. The implementation and tests are a bit 
different, but I ran your test cases and they pass as well.

Thanks a lot!
```

Original comment by `thomas.t...@gmail.com` on 22 Nov 2011 at 7:28
- Changed state: **Accepted**

**Comment by GoogleCodeExporter** [comment]
```
Fixed in version 1.3.162
```

Original comment by `thomas.t...@gmail.com` on 26 Nov 2011 at 12:55
- Changed state: **Fixed**
